### PR TITLE
ARROW-14495: [Python] Fix DictionaryArray.from_buffers, should not crash

### PR DESCRIPTION
**Pull request by milesgranger**
Fix [ARROW-14495](https://issues.apache.org/jira/browse/ARROW-14495)

### CODE REVIEW

File: python/pyarrow/array.pxi [hunk 1]
```diff
@@ -2478,6 +2478,55 @@ cdef class DictionaryArray(Array):
...
+            shared_ptr[CArrayData] c_data
+            shared_ptr[CArray] c_result
+
+        for buf in buffers:
+            c_buffers.push_back(pyarrow_unwrap_buffer(buf))
+
+        if isinstance(dictionary, Array):
+            _dictionary = dictionary
+        else:
+            _dictionary = array(dictionary, memory_pool=memory_pool)
```
Comment (jorisvandenbossche) [inline python/pyarrow/array.pxi:2516]: Given that this is a "power-user" method anyway, I think it is maybe fine to require an Array here for the dictionary (for the Buffers we also only accept actual buffers, and that would avoid needing to add the `memory_pool` keyword, and avoid having to test this, which isn't the case at the moment)

File: python/pyarrow/array.pxi [hunk 2]
```diff
@@ -2478,6 +2478,55 @@ cdef class DictionaryArray(Array):
 
         return self._indices
 
+    @staticmethod
+    def from_buffers(DataType type, int64_t length, buffers, dictionary,
+                     int64_t null_count=-1, int64_t offset=0, memory_pool=None):
+        """
+        Construct a DictionaryArray from buffers.
+
+        type : pyarrow.DataType
```
Comment (jorisvandenbossche) [inline python/pyarrow/array.pxi:2487]:
```suggestion
        Parameters
        ----------
        type : pyarrow.DataType
```

Comment (jorisvandenbossche) [reply python/pyarrow/array.pxi:2487]: And can you also add a "Returns" section?

File: python/pyarrow/tests/test_array.py [hunk 3]
```diff
@@ -725,6 +725,13 @@ def test_struct_array_from_chunked():
         pa.StructArray.from_arrays([chunked_arr], ["foo"])
 
 
+def test_dictionary_from_buffers():
+    a = pa.array(["one", "two", "three", "two", "one"]).dictionary_encode()
+    b = pa.DictionaryArray.from_buffers(
+        a.type, len(a), a.indices.buffers(), a.dictionary)
+    assert a == b
+
```
Comment (jorisvandenbossche) [inline python/pyarrow/tests/test_array.py:733]: Maybe add a test with an offset as well?

**Review by jorisvandenbossche — COMMENTED** [review]
Looking good!

File: python/pyarrow/array.pxi [hunk 4]
```diff
@@ -2478,6 +2478,53 @@ cdef class DictionaryArray(Array):
...
+
+        for buf in buffers:
+            c_buffers.push_back(pyarrow_unwrap_buffer(buf))
+
+        c_type = pyarrow_unwrap_data_type(type)
+
+        with nogil:
+            c_data = CArrayData.Make(
+                c_type, length, c_buffers, null_count, offset)
+            c_data.get().dictionary = dictionary.sp_array.get().data()
```
Comment (jorisvandenbossche) [inline python/pyarrow/array.pxi:2521]: There is a `ArrayData::Make` variant that directly accepts a dictionary, but that might not be worth exposing for just this one line.

**Review by jorisvandenbossche — APPROVED** [review]
Thanks for the updates!

File: python/pyarrow/array.pxi (hunk 4 above)
Comment (milesgranger) [reply python/pyarrow/array.pxi:2521]: I saw that, and it appears to essentially do the same; creating `ArrayData` then setting the dictionary member. Using it this way, we don't need to pass (null?) child data, albeit, if that way is the preferred method I have no problem switching. 👍

File: python/pyarrow/array.pxi [hunk 5]
```diff
@@ -2478,6 +2478,53 @@ cdef class DictionaryArray(Array):
...
+        """
+        Construct a DictionaryArray from buffers.
+
+        Parameters
+        ----------
+        type : pyarrow.DataType
+        length : int
+            The number of values in the array.
+        buffers : List[Buffer]
+            The buffers backing this array.
```
Comment (pitrou) [inline python/pyarrow/array.pxi:2493]: Would this be less ambiguous?
```suggestion
            The buffers backing the indices array.
```

File: python/pyarrow/array.pxi [hunk 6]
```diff
@@ -2478,6 +2478,53 @@ cdef class DictionaryArray(Array):
...
+        dictionary : pyarrow.Array, ndarray or pandas.Series
+            The array of values referenced by the indices.
+        null_count : int, default -1
+            The number of null entries in the array. Negative value means that
```
Comment (pitrou) [inline python/pyarrow/array.pxi:2497]:
```suggestion
            The number of null entries in the indices array. Negative value means that
```

**Review by pitrou — COMMENTED** [review]
Thanks for doing this. Just a minor suggestion.

**Comment by milesgranger** [comment]
@pitrou I think this is ready for another round. :)

**Review by pitrou — APPROVED** [review]
LGTM, thank you @milesgranger

**Pull request merged**